File: src/Neuralm.Presentation.Web/src/messaging/MessageConstructor.ts
import { IMessageSerializer } from '@/interfaces/IMessageSerializer';
import { IMessage } from '@/interfaces/IMessage';
import messageTypeCache from './MessageTypeCache';
import { Message } from './Message';

/**
 * Represents the MessageConstructor class.
 */
export default class MessageConstructor {
  private readonly _textEncoder: TextEncoder = new TextEncoder();
  private readonly _messageSerializer: IMessageSerializer;

  /**
   * Initializes a new instance of the MessageConstructor class.
   */
  constructor(messageSerializer: IMessageSerializer) {
    this._messageSerializer = messageSerializer;
  }

  public constructMessage(message: IMessage): Message {
    const body: Uint8Array = this._messageSerializer.serialize(message);
    const bodySizeBytes: Uint8Array = new Uint8Array(4);
    let dv: DataView = new DataView(bodySizeBytes.buffer);
    dv.setInt32(0, body.length, true);
    const typeName: string = messageTypeCache.getMessageType(message.constructor.name);
    const typeNameBytes: Uint8Array = this._textEncoder.encode(typeName);
    const headerSizeBytes: Uint8Array = new Uint8Array(4);
    dv = new DataView(headerSizeBytes.buffer);
    dv.setInt32(0, 8 + typeNameBytes.length, true);
    const header: Uint8Array = this.concat(this.concat(headerSizeBytes, bodySizeBytes), typeNameBytes);
    return new Message(header, body);
  }

  private concat(a: Uint8Array, b: Uint8Array): Uint8Array {
    const tmp = new Uint8Array(a.length + b.length);
    tmp.set(a, 0);
    tmp.set(b, a.length);
    return tmp;
  }
}

File: src/Neuralm.Services/Neuralm.Services.Common.Infrastructure/Messaging/MessageConstructor.cs
using Neuralm.Services.Common.Application.Interfaces;
using System;
using System.Linq;
using System.Text;

namespace Neuralm.Services.Common.Infrastructure.Messaging
{
    /// <summary>
    /// Represents the <see cref="MessageConstructor"/> class.
    /// </summary>
    internal class MessageConstructor
    {
        private readonly IMessageSerializer _messageSerializer;

        /// <summary>
        /// Initializes an instance of the <see cref="MessageConstructor"/> class.
        /// </summary>
        /// <param name="messageSerializer"></param>
        internal MessageConstructor(IMessageSerializer messageSerializer)
        {
            _messageSerializer = messageSerializer ?? throw new ArgumentNullException(nameof(messageSerializer));
        }

        /// <summary>
        /// Constructs a sendable message from an object.
        /// </summary>
        /// <param name="messageBody">The object.</param>
        /// <returns>Returns a <see cref="Message"/> containing header and body bytes.</returns>
        internal Message ConstructMessage(object messageBody)
        {
            ReadOnlyMemory<byte> body = _messageSerializer.Serialize(messageBody);
            byte[] bodySizeBytes = BitConverter.GetBytes(body.Length);
            byte[] typeNameBytes = Encoding.UTF8.GetBytes(messageBody.GetType().FullName);
            byte[] headerSizeBytes = BitConverter.GetBytes(4 + bodySizeBytes.Length + typeNameBytes.Length);
            ReadOnlyMemory<byte> header = headerSizeBytes.Concat(bodySizeBytes).Concat(typeNameBytes).ToArray();
            return new Message(header, body);
        }

        /// <summary>
        /// Deconstructs a message body as bytes with a given type.
        /// </summary>
        /// <param name="messageBody">The memory body in bytes.</param>
        /// <param name="type">The type.</param>
        /// <returns>Returns the message body as an object.</returns>
        internal object DeconstructMessageBody(Memory<byte> messageBody, Type type)
        {
            return _messageSerializer.Deserialize(messageBody, type);
        }
    }
}
